stage #3/java/module #2. REST API Basics/task02/service/src/main/java/by/epam/esm/service/impl/GiftCertificateServiceImpl.java

```diff
@@ -0,0 +1,204 @@
+import by.epam.esm.dto.mapper.GiftCertificateMapper;
+import by.epam.esm.dto.mapper.GiftCertificateRequestMapper;
+import by.epam.esm.dto.mapper.PaginationMapper;
+import by.epam.esm.enity.GiftCertificate;
+import by.epam.esm.enity.Pagination;
+import by.epam.esm.enity.request.GiftCertificateRequestParam;
+import by.epam.esm.exception.ErrorCode;
+import by.epam.esm.exception.ErrorMessage;
+import by.epam.esm.exception.ServiceException;
+import by.epam.esm.service.GiftCertificateService;
+import by.epam.esm.service.TagService;
+import by.epam.esm.validator.BaseValidator;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+@Service
+public class GiftCertificateServiceImpl implements GiftCertificateService {
+
+    private GiftCertificateMapper giftCertificateMapper;
+    private CrudGiftCertificateDao giftCertificateDao;
+    private BaseValidator baseValidator;
+    private PaginationMapper paginationMapper;
+    private GiftCertificateRequestMapper giftCertificateRequestMapper;
+    private TagService tagService;
+
+    @Autowired
+    public GiftCertificateServiceImpl(GiftCertificateMapper giftCertificateMapper,
+                                      CrudGiftCertificateDao giftCertificateDao,
+                                      BaseValidator baseValidator,
+                                      PaginationMapper paginationMapper,
+                                      GiftCertificateRequestMapper giftCertificateRequestMapper,
+                                      TagService tagService) {
+        this.giftCertificateMapper = giftCertificateMapper;
+        this.giftCertificateDao = giftCertificateDao;
+        this.baseValidator = baseValidator;
+        this.paginationMapper = paginationMapper;
+        this.giftCertificateRequestMapper = giftCertificateRequestMapper;
+        this.tagService = tagService;
+    }
+
+    @Override
+    public GiftCertificateDto findById(Integer id) {
+        GiftCertificateDto giftCertificateDto = giftCertificateDao.findById(id).stream()
+                .findAny()
+                .map(giftCertificateMapper::giftCertificateToGiftCertificateDto)
+                .orElseThrow(() -> new ServiceException(
+                        ErrorCode.NOT_FIND_CERTIFICATE_BY_ID,
+                        ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage(),
+                        Set.of(new ErrorMessage("id", Integer.toString(id), ErrorCode.NOT_FIND_CERTIFICATE_BY_ID.getMessage()))
+                ));
+        return setTagsForDto(giftCertificateDto);
+    }
+
+    private GiftCertificateDto setTagsForDto(GiftCertificateDto giftCertificateDto) {
+        giftCertificateDto.setTags(tagService.findByGiftCertificateId(giftCertificateDto.getId()));
+        return giftCertificateDto;
+
+    }
+
+    @Override
+    public GiftCertificateDto add(GiftCertificateDto dto) {
```

Comment:
add or save? in one place you have save(), in another add(), please follow the same naming strategy